Sources/FluentData/FluentDataContext.swift

```diff
@@ -331,77 +330,168 @@ extension FluentDataContext: DatabaseStateTracker {
+            guard let applicationSupportDirectory = try? FileManager.default.url(
+                for: .applicationSupportDirectory,
+                in: .userDomainMask,
+                appropriateFor: nil,
+                create: true
+            ) else {
+                throw FluentDataContextError.unknownPathToDatabaseFile
             }
 
             guard let urlSafePersistenceName = name.addingPercentEncoding(
                 withAllowedCharacters: CharacterSet.urlPathAllowed.subtracting(CharacterSet.symbols).subtracting(CharacterSet.newlines)
             ) else {
-                fatalError("Unknown to store database `\(name)` (2)")
+                throw FluentDataContextError.invalidDatabaseName
             }
 
-            let filePath = folder.appendingPathComponent("\(urlSafePersistenceName).sqlite")
+            let filePath = applicationSupportDirectory.appendingPathComponent("\(urlSafePersistenceName).sqlite")
+            return filePath
+
+        case .iCloud(let container, let name):
+            guard let iCloudFolder = FileManager.default.url(forUbiquityContainerIdentifier: container) else {
+                return try Self.localFallbackFilePathUnsafe()
+            }
+
+            guard let urlSafePersistenceName = name.addingPercentEncoding(
+                withAllowedCharacters: CharacterSet.urlPathAllowed.subtracting(CharacterSet.symbols).subtracting(CharacterSet.newlines)
+            ) else {
+                throw FluentDataContextError.invalidDatabaseName
+            }
+
+            let filePath = iCloudFolder.appendingPathComponent("\(urlSafePersistenceName).sqlite")
             return filePath
 
         case .memory:
             return nil
         }
     }
 
-    static var removableFilePath: URL? {
+    static func filePathUnsafe() throws -> URL {
+        guard let path = try filePath() else {
+            fatalError("filePathUnsafe called for an unsupported persistence type")
+        }
+        return path
+    }
+
+    static func localFallbackFilePathUnsafe() throws -> URL {
         switch Self.persistence {
-        case .bundle, .memory:
-            return nil
+        case .iCloud(let container, let name):
+            guard let applicationSupportDirectory = try? FileManager.default.url(
+                for: .applicationSupportDirectory,
+                in: .userDomainMask,
+                appropriateFor: nil,
+                create: true
+            ) else {
+                throw FluentDataContextError.unknownPathToDatabaseFile
+            }
 
-        case .file:
+            let iCloudLocalDirectory = applicationSupportDirectory.appendingPathComponent("iCloudLocal").appendingPathComponent(container)
+
+            guard let urlSafePersistenceName = name.addingPercentEncoding(
+                withAllowedCharacters: CharacterSet.urlPathAllowed.subtracting(CharacterSet.symbols).subtracting(CharacterSet.newlines)
```

Comment:
Can be a util extension?

Reply:
Relatively specific so I keep it as a (file)private extension but I agree, it makes the code cleaner.